### ISSUE TITLE
Very slow after migrating to 7.0

### ISSUE THREAD

**Issue by kletellier** [ampache/ampache#4053]
### Describe the bug

I've a private instance hosted on a OVH web hosting (Apache with MySQL 8 also hosted on ovh), all working fine on 6.6 but after migrating to 7.0 all request are very slow (it take few minutes to displaing index page...)

### To reproduce

Migrate from 6.6 to 7.0 on a web hosted environnment OVH Web 
 

## Environment

* Ampache version: 7.0.0
* Web server + version: Apache (unknown version) PHP 8.2 or 8.3 jhave the same effect
* Server operating system: Linux Debian GNU/Linux 10 (buster)
* Client operating system: Firefox, Dsub 

### Client type

* Web Firefox
* Ampache 7.0
* SubSonic DSUB

## Settings
 
No specific settings

i've a big db with many thousands of titles

## Logs
 
```
2024-10-24T09:14:55+02:00 [admin] (Ampache\Module\System\Session) -> Writing to a35ee952d2044c3c267e941b39bd8aa1 with expiration 1729757695
2024-10-24T09:14:56+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:14:56+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:14:56+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:14:56 +0200 extension length 3600
2024-10-24T09:14:56+02:00 [admin] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Utility\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:15:00 +0200 extension length 3600
2024-10-24T09:15:00+02:00 [admin] (ajax.server) -> Called for page: {index}
2024-10-24T09:15:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $limit declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
2024-10-24T09:15:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $offset declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
2024-10-24T09:15:00+02:00 [admin] (Ampache\Module\System\Session) -> Writing to a35ee952d2044c3c267e941b39bd8aa1 with expiration 1729757700
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:00+02:00 [ampache] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
2024-10-24T09:15:08+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:08+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:09+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:15:09 +0200 extension length 3600
2024-10-24T09:15:09+02:00 [admin] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Admin\System\WriteConfigAction" for action "write_config"
2024-10-24T09:15:09+02:00 [admin] (Ampache\Module\System\Session) -> Writing to a35ee952d2044c3c267e941b39bd8aa1 with expiration 1729757709
2024-10-24T09:15:09+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:09+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:15:09+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:15:09 +0200 extension length 3600
2024-10-24T09:15:09+02:00 [admin] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Index\ShowAction" for action "show"
2024-10-24T09:16:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:16:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:16:00+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:16:00 +0200 extension length 3600
2024-10-24T09:16:00+02:00 [admin] (ajax.server) -> Called for page: {index}
2024-10-24T09:16:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $limit declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
2024-10-24T09:16:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $offset declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
2024-10-24T09:16:00+02:00 [admin] (Ampache\Module\System\Session) -> Writing to a35ee952d2044c3c267e941b39bd8aa1 with expiration 1729757760
2024-10-24T09:17:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:17:00+02:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-24T09:17:00+02:00 [admin] (Ampache\Module\System\Session) -> a35ee952d2044c3c267e941b39bd8aa1 has been extended to Thu, 24 Oct 2024 10:17:00 +0200 extension length 3600
2024-10-24T09:17:00+02:00 [admin] (ajax.server) -> Called for page: {index}
2024-10-24T09:17:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $limit declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
2024-10-24T09:17:00+02:00 [ampache] (log.lib) -> [Error] Optional parameter $offset declared before required parameter $includes is implicitly treated as a required parameter in file /home/kletellicl/music.kletellier.xyz/public/vendor/mikealmond/musicbrainz/src/MusicBrainz/MusicBrainz.php(492)
```

**Comment by lachlan-00** [comment]
a couple of people have mentioned this today.

minutes is definitely not right but i haven't changed a lot in the backend php that would be obvious.

i'm not here today but will start working on it tomorrow as even the demo site is a bit slower than it should be

**Comment by lachlan-00** [comment]
I think it might be a lack of indexing on user_preference table which has had the biggest changes

**Comment by kletellier** [comment]
don't hesitate to ask me if you need to test anything

**Comment by fufroma** [comment]
I created this index one year ago

`CREATE INDEX user_preference_preference_user_idx  ON user_preference (user, preference);`

**Comment by lachlan-00** [comment]
I took out the preference rebuilds and that's sped up the develop sites.

need to look into more. it would be good to see if anything gets stuck in the sql with `show processlist;`

**Comment by lachlan-00** [comment]
added new indexes, makes a big difference
![image](https://github.com/user-attachments/assets/f7202cd6-8b05-420b-80a9-c39c6af9f36c)

**Comment by lachlan-00** [comment]
caching speeds it up another 10ish% but that should help

**Comment by kletellier** [comment]
i've change nothing since 7.0 and processlist seems to be clear, it's not a db issue imo 
![image](https://github.com/user-attachments/assets/0365284f-648a-4aff-8a51-609dc8a0e8b1)

**Comment by lachlan-00** [comment]
i think the recently played calls are slow, just digging into those now

**Comment by lachlan-00** [comment]
4s between query and results, need to pick that up

**Comment by lachlan-00** [comment]
74c8444cc speeds up a lot of things for the now playing checks.

**Comment by kletellier** [comment]
no change after applying this commit... 

120-140 sec everytime for loading start page  (and after Ctrl-F5)

here is the log after cold start (first start of the day, very strange we have the major lag time bfore 

(Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Index\ShowAction" for action "show"

`2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Session) -> Unable to read value from key e43fea543373afdcbe937ffc21d32b40 no data found
2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Session) -> Writing to e43fea543373afdcbe937ffc21d32b40 with expiration 1730276625
2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Session) -> Unable to read value from key aaefdf5e45f2d9f1976d3b25603664ec no data found
2024-10-30T08:23:45+01:00 [ampache] (Ampache\Module\System\Session) -> Session created: aaefdf5e45f2d9f1976d3b25603664ec
2024-10-30T08:23:45+01:00 [admin] (Ampache\Module\System\Session) -> Writing to aaefdf5e45f2d9f1976d3b25603664ec with expiration 1730276625
2024-10-30T08:23:45+01:00 [admin] (Ampache\Module\System\Session) -> Writing to aaefdf5e45f2d9f1976d3b25603664ec with expiration 1730276625
2024-10-30T08:23:45+01:00 [admin] (Ampache\Module\System\Session) -> Unable to read value from key 16da02d08301cce21cfe91ce200a1efe no data found
2024-10-30T08:23:45+01:00 [admin] (Ampache\Module\System\Session) -> Session created: 16da02d08301cce21cfe91ce200a1efe
2024-10-30T08:23:45+01:00 [admin] (Ampache\Module\System\Session) -> auth_remember session found
2024-10-30T08:23:46+01:00 [admin] (Ampache\Module\System\Session) -> 16da02d08301cce21cfe91ce200a1efe has been extended to Wed, 30 Oct 2024 09:23:46 +0100 extension length 3600
2024-10-30T08:23:46+01:00 [admin] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Index\ShowAction" for action "show"
2024-10-30T08:25:57+01:00 [admin] (Ampache\Module\System\Session) -> Session created: 58b9ce1beb2a36e2715d0dbf17b0a1d3`

![image](https://github.com/user-attachments/assets/4e91ec64-8845-4f48-98eb-87b48f7b44ae)

**Comment by kletellier** [comment]
it's not related to the catalog size, i've setup a blank instance (7  8.3 squashed version) always on my ovh web env and the db in mariadb on remote vps 
i've the same issue near 2 mn to display the index page

![image](https://github.com/user-attachments/assets/9b6cda88-4995-439f-93a8-21c3a487372d)

**Comment by lachlan-00** [comment]
yeah catalog size won't be an issue in general it's all mapped and indexed. 2mins is impressively long. i'll do a squashed install locally and see if there's something missing in that branch

**Comment by lachlan-00** [comment]
on your tester can you disable all the home plugins
![image](https://github.com/user-attachments/assets/d7a217bd-df2b-4647-9022-d7d18cba88f5)

and turn off recently played in the interface sections

"show now playing" and "Show Albums of the Moment"

**Comment by kletellier** [comment]
same issue :-(
![image](https://github.com/user-attachments/assets/4f25100e-5792-4dfa-a6bf-d4fe02254586)

**Comment by fufroma** [comment]
Huge install here, just upgraded to 7.0.1

A bit slower than before, but not like you @kletellier 

Feel free to test on https://play.dogmazic.net/

But I do have a slower cron (from 3m to 4m, https://github.com/ampache/ampache/issues/3970 would have been useful)

**Comment by lachlan-00** [comment]
okay so this must be a different issue than what we've been finding. it could be server related as this is every page?

is it also slow loading between sidebar tabs the same way?

**Comment by fufroma** [comment]
There are some query that could explain some impact.

```
SELECT COUNT(`name`) from `user_preference`;
1 row in set (1,044 sec)
=> ok I have 1063360 items in this table...

SELECT `object_id` AS `id`, MAX(`count`) AS `count` FROM `cache_object_count` WHERE `object_type` = 'album' AND `count_type` = 'stream' AND `threshold` = '365' GROUP BY `object_id`, `object_type` ORDER BY `count` DESC  LIMIT 10 ;
[...]
=> this one may took as 3second on first load, then as the query is cached, only 0.2s... and after a moment 3s again

```

I think some clean from Ampache 7 added some SQL query... and as my db is huge, it add time in things you didn't expect (for example, I have more than 8k users... so user_preference is pretty huge).

**Comment by fufroma** [comment]
There are definitely something to do on the sql.

If I replace `Ampache\Module\Statistics\Stats::get_top` by a dumb `return 1;`, the page https://play.dogmazic.net/mashup.php?action=album go from 2.5s to 0.8s (it may even be 4s if the query aren't cached)

**Comment by fufroma** [comment]
Mashup page do query on `object_count` table instead of `cache_object_count`, don't know if that's expected.

**Comment by fufroma** [comment]
https://github.com/ampache/ampache/pull/4061 got me a few hundreds ms quicker

**Comment by kletellier** [comment]
> okay so this must be a different issue than what we've been finding. it could be server related as this is every page?
> 
> is it also slow loading between sidebar tabs the same way?

Yes every click take nearly 2mn...

**Comment by fufroma** [comment]
@kletellier did you enable `memory_cache` and are the cron jobs running fine?

**Comment by kletellier** [comment]
yes same effect with memory cache on or off, what do you mean about cron jobs ? i've no cron jobs in my previous versions.
This issue looks like  timeout issue but i can't see where...

**Comment by fufroma** [comment]
take a look a that https://github.com/ampache/ampache/wiki/cron

**Comment by fufroma** [comment]
Any improvement?

**Comment by kletellier** [comment]
sorry i've not replied, my instance are hosted on a web public env in ovh (french cloud provider), i've not found the way to implement cron in this environnment i'll check it this week :)

**Comment by kletellier** [comment]
i've tested by running manually the cron with the CLI for caching queries, no more update, always 130 sec for each page displaying....

**Comment by dmersiowsky** [comment]
Same here after upgrading from 6.x to 7.0.1 today. 
Response times went from less than a second to this:

<img width="307" alt="Screenshot 2024-12-13 at 12 41 58" src="https://github.com/user-attachments/assets/14c32dda-6886-4a20-8640-d7a6809638ae" />

Cache on/off and cron don't make a difference. I have a library with >10K titles.

**Comment by fufroma** [comment]
@dmersiowsky were you able to pinpoint what's taking so long? There are less query and more cache, so there is something happening here. And we aren't able to reproduce it on all Ampache setup (my instance got 61k songs, there are some performance issue but not that big).

**Comment by lachlan-00** [comment]
there were a lot of things slowing down that i hadn't noticed but my index page in 7.1.0 will load in 0.89s compared to 4+ on 7.0.1.

I would disable and plugins with the category 'Home'.

we looked at the process list so it wasn't the sql side but is there a hard memory limit on your webserver and 7 is using more ram than before?

**Comment by lachlan-00** [comment]
One thing i've noticed a few times is memory leaks in areas like search/browse when there are mistakes in the code.

One possible issue is a process starts and never completes locking up the server. which is why the sql is clean

**Comment by fufroma** [comment]
Both people with this issue have similar time (2 minutes).

Could it be a timeout? For example ampache try to contact an external server to check for update but it's firewalled, so it's blocked until it timeout?

**Comment by fufroma** [comment]
Note for myself: to check timeout theory we can't just remove internet access from ampache, network will fail instantly. Must check with DROP firewall rule to have silent fail, so application wait for timeout

**Comment by dmersiowsky** [comment]
> @dmersiowsky were you able to pinpoint what's taking so long? There are less query and more cache, so there is something happening here. And we aren't able to reproduce it on all Ampache setup (my instance got 61k songs, there are some performance issue but not that big).

Did not have the time to investigate deeper, but will try to come up with something.
"Less query and more cache" is because I had cache disabled with the old setup. With caching enabled it looks like this:

**Queries: 308 | Cache Hits: 61 | Load Time: 0.0358**

As for @fufroma's theory: I do not limit/block outgoing traffic. If it's a timeout issue, it's certainly not due to a firewall blocking traffic. But it your assumption makes sense, the quite constant times ~120s are really somewhat suspicious.

What I can say: 

- It's not like the system shows high CPU load or anything. Server activity is quite low overall. So it indeed seems like some kind of waiting time / timeout.
- MariaDB slow queries log is not showing anything
- It's not only slow for catalog-related requests. Even e.g. opening (or even saving) the settings shows these high response times

**Comment by dmersiowsky** [comment]
Think I found it. It's a cURL to localhost in `Module/Util/Environment.php#isDevJS()`:

`$handle = curl_init('http://localhost:5177/' . $entry);`

My firewall is not responding to requests to this port.

**Comment by kletellier** [comment]
Thank you Daniel, by returning false directly on this method this solve my problem !

**Comment by fufroma** [comment]
@dmersiowsky great find!

**Comment by lachlan-00** [comment]
@mitchray can you tell me what it does? should i put that in the config as an option?

**Comment by lachlan-00** [comment]
I can also look at putting a timeout in which fail over faster
```
public function isDevJS(string $entry): bool
    {
        // hardcoded port for simplicity
        $handle = curl_init('http://localhost:5177/' . $entry, CURLOPT_TIMEOUT, 1);

        if ($handle === false) {
            return false;
        }
```

**Comment by lachlan-00** [comment]
![image](https://github.com/user-attachments/assets/4d1cdc10-cd44-4294-92ba-b57479782d9c)

confirming that's definitely the issue

**Comment by fufroma** [comment]
Why not a parameters in config.php to define if it's a dev or a prod setup?

**Comment by lachlan-00** [comment]
it's in the header template so a config option is good enough

**Comment by lachlan-00** [comment]
set after db5fa0fc9cea52bfc1b020af551a93b3e09afaf0

**Comment by fufroma** [comment]
7.1.0 is released with the fix, I think we can close this issue

**Comment by kletellier** [comment]
Upgraded also on this relase and now everything works fine and faster.
Great work !
Thank you

**Comment by kletellier** [comment]
Fixed with 7.1
